plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/TransactionScopeCachingDirectoryLister.java

```diff
@@ -0,0 +1,230 @@
+        }
+        delegate.invalidate(partition);
+    }
+
+    private RemoteIterator<LocatedFileStatus> cachingRemoteIterator(FetchingValueHolder cachedValueHolder, Path path)
+    {
+        return new RemoteIterator<>()
+        {
+            private int fileIndex;
+
+            @Override
+            public boolean hasNext()
+                    throws IOException
+            {
+                try {
+                    boolean hasNext = cachedValueHolder.getCachedFile(fileIndex).isPresent();
+                    // Update cache weight of cachedValueHolder for a given path.
+                    // The cachedValueHolder acts as an invalidation guard. If a cache invalidation happens while this iterator goes over
+                    // the files from the specified path, the eventually outdated file listing will not be added anymore to the cache.
+                    cache.asMap().replace(path, cachedValueHolder, cachedValueHolder);
+                    return hasNext;
+                }
+                catch (IOException exception) {
+                    // invalidate cached value to force retry of directory listing
+                    cache.invalidate(path);
+                    throw exception;
+                }
+            }
+
+            @Override
+            public LocatedFileStatus next()
+                    throws IOException
+            {
+                // force cache entry weight update in case next file is cached
+                checkState(hasNext());
+                return cachedValueHolder.getCachedFile(fileIndex++).orElseThrow();
+            }
+        };
+    }
+
+    @VisibleForTesting
+    boolean isCached(Path path)
+    {
+        FetchingValueHolder cached = cache.getIfPresent(path);
+        return cached != null && cached.isFullyCached();
+    }
+
+    private static boolean isLocationPresent(Storage storage)
+    {
+        // Some Hive table types (e.g.: views) do not have a storage location
+        return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
+    }
+
+    private static class FetchingValueHolder
+    {
+        private final List<LocatedFileStatus> cachedFiles = synchronizedList(new ArrayList<>());
+        @GuardedBy("this")
+        @Nullable
+        private RemoteIterator<LocatedFileStatus> fileIterator;
+        @GuardedBy("this")
+        @Nullable
+        private IOException exception;
+
+        public FetchingValueHolder(RemoteIterator<LocatedFileStatus> fileIterator)
+        {
+            this.fileIterator = requireNonNull(fileIterator, "fileIterator is null");
+        }
+
+        public synchronized boolean isFullyCached()
+        {
+            return fileIterator == null && exception == null;
+        }
+
+        public int getCachedFilesSize()
+        {
+            return cachedFiles.size();
+        }
+
+        public RemoteIterator<LocatedFileStatus> getCachedFiles()
```

Comment:
the signature of this method -- get cached files returning Remote iterator -- looks weird.
it would be better to return `List`, or plain `Iterator` , and let the caller convert this into a RemoteIterator